tests/acme_endorsement.py

```diff
@@ -0,0 +1,341 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the Apache 2.0 License.
+import subprocess
+import os
+import time
+import urllib.request
+import json
+
+import infra.network
+import infra.path
+import infra.proc
+import infra.interfaces
+import infra.net
+import infra.e2e_args
+import infra.crypto
+import suite.test_requirements as reqs
+from socket import socket
+import ssl
+from cryptography.x509 import load_der_x509_certificate, load_pem_x509_certificate
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat
+
+from loguru import logger as LOG
+
+
+def get_network_public_key(network):
+    cert_data = open(os.path.join(network.common_dir, "service_cert.pem"), "rb").read()
+    network_cert = load_pem_x509_certificate(cert_data, default_backend())
+    return network_cert.public_key().public_bytes(
+        encoding=Encoding.PEM, format=PublicFormat.SubjectPublicKeyInfo
+    )
+
+
+@reqs.description("Start network and wait for ACME certificates")
+def wait_for_certificates(args, network_name, ca_certs, timeout=5 * 60):
+    with infra.network.network(
+        args.nodes, args.binary_dir, args.debug_nodes, args.perf_nodes, pdb=args.pdb
+    ) as network:
+        network.start_and_open(args)
+
+        # NB: Pebble non-deterministically injects delays and failures,
+        # so the following checks may take a signficant amount of time.
+
+        network_public_key = get_network_public_key(network)
+
+        start_time = time.time()
+        end_time = start_time + timeout
+        num_ok = 0
+        while num_ok != len(args.nodes):
+            if time.time() > end_time:
+                raise TimeoutError(
+                    f"Not all nodes had the correct ACME-endorsed TLS certificate installed after {timeout} seconds"
+                )
+
+            num_ok = 0
+            for node in network.nodes:
+                iface = node.host.rpc_interfaces["acme_endorsed_interface"]
+                try:
+                    context = ssl.SSLContext(ssl.PROTOCOL_TLSv1_2)
```

Comment:
Why not 1.3?

Reply:
Never occurred to me to try that, I'll switch it and see whether anything breaks.

Reply:
Looks like that's not supported by some of the Python things. It just hangs when I try to use `PROTOCOL_TLSv1_3` without providing any information. Interestingly, it also just hangs when I use `PROTOCOL_TLSv1_5`... However, pylint does report that they both are not available.

Reply:
Very odd, I thought it was just a thin layer on top of openssl. Anyway, quite orthogonal to the feature being added :)